### ISSUE TITLE
Support for named template strings

### ISSUE THREAD

**Issue by ispringle** [guillaumebrunerie/ultimate-js-mode#2]
Some frameworks, like lit, use template strings to define HTML and CSS. For example:
```javascript
const element = html`<div class="foo">This is my new custom element!</div>`
```

Wondering what it would take to add support for that to this mode. Neovim is using treesitter to pull this off, they do it with the following query:
```
(call_expression
  function: (identifier) @injection.language
  arguments: [
    (arguments
      (template_string) @injection.content)
    (template_string) @injection.content
  ])
```

I've tested this in the treesitter playground and it highlights the correct portion of the above JS example. I'm not sure how to turn that query into its own set of nested highlighting though or else I'd make a PR

**Comment by guillaumebrunerie** [comment]
Hey!
I'd love to have that as well (I use styled components in React a lot myself, and it would be great to have CSS syntax highlighting there), but I think that the last time I investigated it, my conclusion was that Emacs has no support whatsoever for injection queries.
I might be wrong, or it might have changed, maybe I'll try again to make it work.

**Comment by guillaumebrunerie** [comment]
I got something partially working, but then ran into some Emacs bugs (for instance [this one](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=73324)), so I'll wait until Emacs 30 to try again.

**Comment by ispringle** [comment]
@guillaumebrunerie is it in a branch? I have the latest version of emacs-30 running, I can test it for you. 

I can't find the associated commits that Yuan made to fix this issue, if it's a TS issue that should eventually get backported (I think) to the TS package and than you could manually pull that into your version of emacs, assuming 28 or 29, and that should allow you to continue working with the latest version of TS.

**Comment by guillaumebrunerie** [comment]
It's in the main branch, you should just need to uncomment the following lines to try it out https://github.com/guillaumebrunerie/ultimate-js-mode/blob/dcf39e26193c5fbee3e948574a229eb414bacafe/ultimate-js-mode.el#L194-L203

For now it's very much a proof of concept, it's hard coded (in several places) to check for a `css` tag and has some very basic CSS highlighting queries (maybe it should use the ones bundled in Emacs instead, assuming the grammar didn't change too much).

The main issue I had with it was that even though adding a `css` template string correctly highlighted it, removing the `css` part didn't revert the highlight, so the highlight could easily get completely broken. I believe it was fixed by this commit: https://github.com/emacs-mirror/emacs/commit/0834106a6209833ab5f5c032cce857f39889d3d6, at least the description seems to match what I observed.

**Comment by guillaumebrunerie** [comment]
Actually, there are also three other lines that need to be uncommented:
https://github.com/guillaumebrunerie/ultimate-js-mode/blob/dcf39e26193c5fbee3e948574a229eb414bacafe/ultimate-js-mode.el#L111 and the two similar ones below.

I now quickly tried in Emacs 30, and I can confirm that the issue I mentioned above is fixed, but I noticed a few more issues as well, for instance I’m using CSS faces that are only available after loading the regular `css-mode`, and Emacs 30 complains about not finding some functions (not sure what this is about). Finally I would like embedded languages to have a slightly different background, otherwise it seems to me like it would be very confusing, but this will most likely require a completely different implementation using overlays or so.

Have you actually been using `ultimate-js-mode`? I have mostly designed it for myself without thinking anyone else would use it, so it’s very likely that some parts of it don’t really work well with a different setup.

**Comment by ispringle** [comment]
Yes I have been using ultimate-js-mode for editing UX code because it's handling Lit better than anything else. I use the treesitter version for server-side stuff but for the UX I'm using ultimate-js-mode. I haven't tried your changes yet. I switched to guix last week to evaluate it and haven't figured out how to get emacs 30 installed yet.

**Comment by ispringle** [comment]
Just wanted to follow up. I decided to tackle this all over again. In Emacs 30+ I have a working minor mode that augments the existing js-ts-mode and typescript-ts-mode. It provides syntax highlighting for template strings using teh html-ts and css-ts parsers. It also uses them to figure out the appropriate indentation based on the indent level of the JS it's called from.

https://github.com/ispringle/template-literals-ts-mode
